tensorflow/core/kernels/svd_op_gpu.cu.cc

```diff
@@ -0,0 +1,455 @@
+      if (!full_status.ok()) {
+        full_status.Update(errors::InvalidArgument(kErrMsg));
+      }
+      OP_REQUIRES_OK_ASYNC(context, full_status, done);
+      done();
+    };
+
+    OP_REQUIRES_OK_ASYNC(context, solver.CopyLapackInfoToHostAsync(
+                                      dev_info, std::move(info_checker)),
+                         done);
+  }
+
+  // The SVD if m >= n
+  void PerformSVD_MgeqN(OpKernelContext* context, DoneCallback done, int64 m,
+                        int64 n, int64 p, const gtl::ArraySlice<int32>& perm,
+                        const Tensor& M, Tensor* S, Tensor* U, Tensor* V) {
+    TensorShape shapeRaw = M.shape();
+    shapeRaw.RemoveDim(shapeRaw.dims() - 1);
+    shapeRaw.RemoveDim(shapeRaw.dims() - 1);
+
+    // Transpose M, because cuSolver expects it to be column-major
+    TensorShape input_shape = shapeRaw;
+    input_shape.AddDim(n);
+    input_shape.AddDim(m);
+    Tensor input_copy;
+    OP_REQUIRES_OK_ASYNC(
+        context, context->allocate_temp(M.dtype(), input_shape, &input_copy),
+        done);
+    auto device = context->eigen_device<GPUDevice>();
+    OP_REQUIRES_OK_ASYNC(
+        context, 
+        DoTranspose(device, M, perm, &input_copy),
+        done);
+
+    // I need to transpose U at the end
+    // Not V, because cuSolver work column-major
+    Tensor u_copy;
+    if (compute_uv_) {
+      TensorShape u_shape;
+      if (full_matrices_) {
+        u_shape = U->shape();
+      } else {
+        u_shape = shapeRaw;
+        u_shape.AddDim(p);
+        u_shape.AddDim(m);
+      }
+      OP_REQUIRES_OK_ASYNC(
+          context, context->allocate_temp(U->dtype(), u_shape, &u_copy), done);
+    }
+
+    // get the pointers to the data
+    Scalar* input_ptr;
+    SScalar* outputS_ptr;
+    Scalar* outputU_ptr = NULL;
+    Scalar* outputV_ptr = NULL;
+    auto input_reshaped = input_copy.template flat_inner_dims<Scalar, 3>();
+    input_ptr = input_reshaped.data();
+    outputS_ptr = S->template flat_inner_dims<SScalar, 2>().data();
+    if (compute_uv_) {
+      outputU_ptr = u_copy.template flat_inner_dims<Scalar, 3>().data();
+      outputV_ptr = V->template flat_inner_dims<Scalar, 3>().data();
+    }
+
+    // call the SVD
+    const int64 batch_size = input_reshaped.dimension(0);
+    std::vector<DeviceLapackInfo> dev_info;
+    dev_info.emplace_back(context, batch_size, "gesvd");
+    CudaSolver solver(context);
+    RunSVD(context, done, m, n, p, batch_size, input_ptr, outputS_ptr,
+           outputU_ptr, outputV_ptr, dev_info.back().mutable_data(), solver);
+
+    // Transpose U
+    if (compute_uv_) {
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          DoTranspose(device, u_copy, perm, U),
+          done);
+    }
+
+    // now check if the SVD operation succeeded or not
+    CheckResult(context, done, dev_info, solver, input_copy, u_copy);
+  }
+
+  // The SVD if m < n
+  void PerformSVD_MlessN(OpKernelContext* context, DoneCallback done, int64 m,
```

Comment:
It would seem that you could avoid a lot of duplicated boilerplate code, e.g., by introducing variables "small_dim" and "large_dim" and combining  PerformSVD_MlessN and  PerformSVD_MgeqN.

Reply:
Yes, I know that there is quite some duplication. But because I have to handle the matrices differently in terms of transposition each time I need this. It could have been combined into one function with if-statements on which matrix should be transposed.
There's a tradeoff between two functions for the two cases and one function for both with many if-statements for which matrix to transpose (and whether we need an additional copy of the input matrix).
I doubt that it is worth it.

Reply:
Perhaps put a TODO.